src/muse/carbon_budget.py

```diff
@@ -385,104 +386,94 @@ def bisection(
+        # Cap prices between 0.01 and price_too_high_threshold
+        if refine_price:
+            up = min(up, price_too_high_threshold)
+        low = max(low, 0.01)
 
-    # We create a 2-price sample being lower and upper bound of bisection
-    lb = bisect_loop(market, sectors, equilibrium, commodities, up0)
-    ub = bisect_loop(market, sectors, equilibrium, commodities, low0)
+        # Round prices to 2dp
+        low = round(low, 2)
+        up = round(up, 2)
 
-    # Start bisection loop over a number of iterations
-    # sample_size is in this method used to determine
-    # the number of iterations
-    niter = sample_size
-    for n in range(niter):
-        if refine_price:
-            # We apply a cap if carbon price beyond threshold
-            if max(sample_prices) > price_too_high_threshold:
-                price_too_high_threshold = round(
-                    price_too_high_threshold * (1 + 0.1) ** time_exp, 7
-                )
-                up0 = round(min(max(sample_prices), price_too_high_threshold), 7)
-                low0 = round(
-                    min(min(sample_prices), price_too_high_threshold, up0 * 0.9), 7
-                )
-                lb = bisect_loop(market, sectors, equilibrium, commodities, up0)
-                ub = bisect_loop(market, sectors, equilibrium, commodities, low0)
-        if low0 == up0:
-            new_price = low0
-            break
-        if lb == threshold:
-            new_price = up0
-            break
-        elif ub == threshold:
-            new_price = low0
-            break
-        else:
-            low, up = min_max_bisect(
-                low0,
-                lb,
-                up0,
-                ub,
-                market,
-                sectors,
-                equilibrium,
-                commodities,
-                threshold,  # type: ignore
+        # Calculate carbon emissions at new bounds
+        if low not in emissions_cache:
+            emissions_cache[low] = bisect_loop(
+                market, sectors, equilibrium, commodities, low
             )
-            # Exit loop if low and upper bound on emissions are close
-            if abs(low - up) <= 0.001:
-                new_price = round((low + up) / 2.0, 7)
-                break
-            # Exit loop if low and upper bound on emissions are close to threshold
-            elif abs(ub - threshold) <= abs(0.1 * threshold):
-                new_price = low
-                break
-            elif abs(lb - threshold) <= abs(0.1 * threshold):
-                new_price = up
+        ub = emissions_cache[low]
+        if up not in emissions_cache:
+            emissions_cache[up] = bisect_loop(
+                market, sectors, equilibrium, commodities, up
+            )
+        lb = emissions_cache[up]
+
+        # Terminate early if many consecutive emissions are the same
+        if len(emissions_cache) >= early_termination_count + 2:
```

Comment:
I'm not entirely sure I follow this.

Every time you have an `up` or `low` that is not in the cache, you add them to the cache. So, the cache grows when `up` and `low` are different. However, this is suppose of break the loop if they are the same (too many times). Then, shouldn't this be checking that the cache is not growing and not that is growing and becoming too large?

Reply:
The main reason I included this was for models where it's impossible to get emissions close to the budget (perhaps low-carbon technologies don't exist that can meet the demand). In this case, the upper bound price will keep being ramped up until a plateau on emissions is reached. If this plateau is not within range of the budget, the price will keep being ramped up until the maximum iterations is reached, so I needed a check in place to prevent this.

`up` and `low` refer to prices, but this is checking whether _emissions_ are the same too many times

Reply:
I'm not sure I follow. Where are you checking that emissions are the same too many times? Here you just check that the emissions cache length is above certain threshold.

Reply:
It's in the three lines below that `if` statement. The `if` statement is just to make sure the cache has enough values to check (and I thought it was reasonable to exclude the starting bounds from the check, hence the +2)

Reply:
Got it.